Mathlib/Algebra/BigOperators/Basic.lean

```diff
@@ -90,43 +90,160 @@ and we therefore choose the level `67`.
+syntax bigOpBinderParenthesized := " (" bigOpBinder ")"
+/-- A list of parenthesized binders -/
+syntax bigOpBinderCollection := bigOpBinderParenthesized+
+/-- A single (unparenthesized) binder, or a list of parenthesized binders -/
+syntax bigOpBinders := bigOpBinderCollection <|> (ppSpace bigOpBinder)
+
+/-- Collects additional binder/Finset pairs for the given `bigOpBinder`.
+Note: this is not extensible at the moment, unlike the usual `bigOpBinder` expansions. -/
+def processBigOpBinder (processed : (Array (Term × Term)))
+    (binder : TSyntax ``bigOpBinder) : MacroM (Array (Term × Term)) :=
+  set_option hygiene false in
+  withRef binder do
+    match binder with
+    | `(bigOpBinder| $x:term) =>
+      match x with
+      | `(($a + $b = $n)) => -- Maybe this is too cute.
+        return processed |>.push (← `(⟨$a, $b⟩), ← `(Finset.Nat.antidiagonal $n))
+      | _ => return processed |>.push (x, ← ``(Finset.univ))
+    | `(bigOpBinder| $x : $t) => return processed |>.push (x, ← ``((Finset.univ : Finset $t)))
+    | `(bigOpBinder| $x ∈ $s) => return processed |>.push (x, ← `(finset% $s))
+    | `(bigOpBinder| $x < $n) => return processed |>.push (x, ← `(Finset.Iio $n))
+    | `(bigOpBinder| $x ≤ $n) => return processed |>.push (x, ← `(Finset.Iic $n))
+    | `(bigOpBinder| $x > $n) => return processed |>.push (x, ← `(Finset.Ioi $n))
+    | `(bigOpBinder| $x ≥ $n) => return processed |>.push (x, ← `(Finset.Ici $n))
+    | _ => Macro.throwUnsupported
+
+/-- Collects the binder/Finset pairs for the given `bigOpBinders`. -/
+def processBigOpBinders (binders : TSyntax ``bigOpBinders) :
+    MacroM (Array (Term × Term)) :=
+  match binders with
+  | `(bigOpBinders| $b:bigOpBinder) => processBigOpBinder #[] b
+  | `(bigOpBinders| $[($bs:bigOpBinder)]*) => bs.foldlM processBigOpBinder #[]
+  | _ => Macro.throwUnsupported
+
+/-- Collect the binderIdents into a `⟨...⟩` expression. -/
+def bigOpBindersPattern (processed : (Array (Term × Term))) :
+    MacroM Term := do
+  let ts := processed.map Prod.fst
+  if ts.size == 1 then
+    return ts[0]!
+  else
+    `(⟨$ts,*⟩)
+
+/-- Collect the terms into a product of sets. -/
+def bigOpBindersProd (processed : (Array (Term × Term))) :
+    MacroM Term := do
+  if processed.isEmpty then
+    `((Finset.univ : Finset Unit))
+  else if processed.size == 1 then
+    return processed[0]!.2
+  else
+    processed.foldrM (fun s p => `(SProd.sprod $(s.2) $p)) processed.back.2
+      (start := processed.size - 1)
+
+/--
+- `∑ x, f x` is notation for `Finset.sum Finset.univ f`. It is the sum of `f x`,
+  where `x` ranges over the finite domain of `f`.
+- `∑ x ∈ s, f x` is notation for `Finset.sum s f`. It is the sum of `f x`,
+  where `x` ranges over the finite set `s` (either a `Finset` or a `Set` with a `Fintype` instance).
+- `∑ x ∈ s with p x, f x` is notation for `Finset.sum (Finset.filter p s) f`.
+- `∑ (x ∈ s) (y ∈ t), f x y` is notation for `Finset.sum (s ×ˢ t) (fun ⟨x, y⟩ ↦ f x y)`.
```

Comment:
Is it a valid alternative to elaborate this into a nested sum? That would automatically handle the case of `t` depending on `x`, right?

Reply:
The thought was that if you wanted a nested sum then `∑ (x ∈ s), ∑ (y ∈ t), f x y` is available. Yaël and Bhavik were telling me about how this sort of behavior is helpful for related notations such as for expectation, so it might be worth being consistent here.

A downside though is that this PR is not supporting dependent products, i.e., you can't do `∑ (x ∈ s) (y ∈ t x), f x y`, and it's not very clear to me how to support them in a low-confusion way.